alternator/executor.cc

```diff
@@ -5133,6 +5133,14 @@ static rjson::value encode_paging_state(const schema& schema, const service::pag
         auto exploded_ck = pos.key().explode();
         auto exploded_ck_it = exploded_ck.begin();
         for (const column_definition& cdef : schema.clustering_key_columns()) {
+            if (exploded_ck_it == exploded_ck.end()) {
+                // clustering key in current page position might be shorter than full column definitions, for example
+                // in case where our position is at range tombstone. In normal operation this is impossible, as DynamoDB
+                // doesn't allow range deletions. DynamoDB api can be used to access some system tables thought,
+                // and those are under cql rules, which allow range tombstones. Here we break if position's clustering key
+                // is shorter than column definitions - this is fine and decoder for paging state will handle this correctly.
+                break;
```

Comment:
Nice catch.

By the way I think there's another way you could have solved this bug that perhaps didn't require so much explanations. You can just to replace the loop index:

Instead of looping over clustering_key_columns() and maintaining a second iterator over exploded_ck, reverse the roles: Have a loop over exploded_ck(), and maintain an iterator over clustering_key_columns. If exploded_ck is never **longer** than clustering_key_columns, it will work correctly and not look surprising or require a comment to explain it.

Reply:
> This is what @michoecho suggested

To be clear, I suggested the same change as @nyh.

> if we have preference for iterating over columns i can update

It's not very important, but I'd say that yes, we do have a preference for iterating over columns instead of types. It's a pattern that occurs in several places in the codebase.